src/read/xcoff/symbol.rs

```diff
@@ -0,0 +1,439 @@
+    XcoffSymbolIterator<'data, 'file, xcoff::FileHeader64, R>;
+
+/// An iterator over the symbols of an `XcoffFile`.
+pub struct XcoffSymbolIterator<'data, 'file, Xcoff, R = &'data [u8]>
+where
+    'data: 'file,
+    Xcoff: FileHeader,
+    R: ReadRef<'data>,
+{
+    pub(crate) file: &'file XcoffFile<'data, Xcoff, R>,
+    pub(super) symbols: &'file SymbolTable<'data, Xcoff, R>,
+    pub(super) index: usize,
+}
+
+impl<'data, 'file, Xcoff: FileHeader, R: ReadRef<'data>> fmt::Debug
+    for XcoffSymbolIterator<'data, 'file, Xcoff, R>
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("XcoffSymbolIterator").finish()
+    }
+}
+
+impl<'data, 'file, Xcoff: FileHeader, R: ReadRef<'data>> Iterator
+    for XcoffSymbolIterator<'data, 'file, Xcoff, R>
+{
+    type Item = XcoffSymbol<'data, 'file, Xcoff, R>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let index = self.index;
+        let symbol = self.symbols.symbols.get(index)?;
+        self.index += 1;
+        Some(XcoffSymbol {
+            file: self.file,
+            symbols: self.symbols,
+            index: SymbolIndex(index),
+            symbol,
+        })
+    }
+}
+
+/// A symbol of an `XcoffFile32`.
+pub type XcoffSymbol32<'data, 'file, R = &'data [u8]> =
+    XcoffSymbol<'data, 'file, xcoff::FileHeader32, R>;
+/// A symbol of an `XcoffFile64`.
+pub type XcoffSymbol64<'data, 'file, R = &'data [u8]> =
+    XcoffSymbol<'data, 'file, xcoff::FileHeader64, R>;
+
+/// A symbol of an `XcoffFile`.
+#[derive(Debug, Clone, Copy)]
+pub struct XcoffSymbol<'data, 'file, Xcoff, R = &'data [u8]>
+where
+    'data: 'file,
+    Xcoff: FileHeader,
+    R: ReadRef<'data>,
+{
+    pub(crate) file: &'file XcoffFile<'data, Xcoff, R>,
+    pub(super) symbols: &'file SymbolTable<'data, Xcoff, R>,
+    pub(super) index: SymbolIndex,
+    pub(super) symbol: &'data Xcoff::Symbol,
+}
+
+impl<'data, 'file, Xcoff: FileHeader, R: ReadRef<'data>> read::private::Sealed
+    for XcoffSymbol<'data, 'file, Xcoff, R>
+{
+}
+
+impl<'data, 'file, Xcoff: FileHeader, R: ReadRef<'data>> ObjectSymbol<'data>
+    for XcoffSymbol<'data, 'file, Xcoff, R>
+{
+    #[inline]
+    fn index(&self) -> SymbolIndex {
+        self.index
+    }
+
+    fn name_bytes(&self) -> Result<&'data [u8]> {
+        self.symbol.name(self.symbols.strings)
+    }
+
+    fn name(&self) -> Result<&'data str> {
+        let name = self.name_bytes()?;
+        str::from_utf8(name)
+            .ok()
+            .read_error("Non UTF-8 XCOFF symbol name")
+    }
+
+    #[inline]
+    fn address(&self) -> u64 {
+        match self.symbol.n_sclass() {
+            xcoff::C_GSYM
+            | xcoff::C_BCOMM
+            | xcoff::C_DECL
+            | xcoff::C_ENTRY
+            | xcoff::C_ESTAT
+            | xcoff::C_ECOMM
+            | xcoff::C_FILE
+            | xcoff::C_BSTAT
+            | xcoff::C_RPSYM
+            | xcoff::C_RSYM => 0,
+            _ => self.symbol.n_value().into(),
```

Comment:
Reverse the logic of this match so that instead we use the list of symbols that we know have addresses: C_EXT, C_WEAKEXT, C_HIDEXT, C_FCN, C_BLOCK, C_STAT